vendor: bump runc to rc94

kolyshkin opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

#### Special notes for your reviewer:

runc v1.0.0-rc94 release notes: https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc94

This also picks up a couple of commits from #100631.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

giuseppe committed:
test: adjust number of expected page faults

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
kubelet: reuse manager

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

kolyshkin committed:
vendor: bump runc to rc94

One notable change is cgroup manager's Set now accept Resources rather
than Cgroup (see https://github.com/opencontainers/runc/pull/2906).
Modify the code accordingly.

Also update runc dependencies (as hinted by hack/lint-depdendencies.sh):

	github.com/cilium/ebpf v0.5.0
	github.com/containerd/console v1.0.2
	github.com/coreos/go-systemd/v22 v22.3.1
	github.com/godbus/dbus/v5 v5.0.4
	github.com/moby/sys/mountinfo v0.4.1
	golang.org/x/sys v0.0.0-20210426230700-d19ff857e887
	github.com/google/go-cmp v0.5.4
	github.com/kr/pretty v0.2.1
	github.com/opencontainers/runtime-spec v1.0.3-0.20210326190908-1c3f411f0417

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

adisky commented:
/ok-to-test
/triage accepted
/sig node

adisky reviewed (commented):
@kolyshkin Thanks for this PR, test failures are indicating you need to run /hack/update-vendor.sh, you can follow this doc for updating dependencies https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md#adding-or-updating-a-dependency

kolyshkin commented:
@adisky yep; updated

kolyshkin committed:
pkg/kubelet/nodeshutdown/systemd: fix for dbus 5.0.4

dbus 5.0.4 adds StoreProperty method which needs to be implemented for
the mock.

Fixes the errors like

> pkg/kubelet/nodeshutdown/systemd/inhibit_linux_test.go:88:9: cannot use f.fakeDBusObject (variable of type *fakeDBusObject) as dbus.BusObject value in return statement: missing method StoreProperty

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

giuseppe reviewed (approved):
LGTM

ehashman commented:
/cc @dims @bobbypage 

I think normally we bump runc by bumping cadvisor rather than on its own?

SergeyKanzhelev commented:
/priority important-soon

marked as "Waiting for author" to address the @ehashman comment

bobbypage commented:
> I think normally we bump runc by bumping cadvisor rather than on its own?

yes, that's normally the course of action. 

I can help cut another cAdvisor release with updated runc, but I don't believe that's critical / has to block the update in k/k. I think we should be able to update runc by itself. @dims is that accurate?

dims commented:
> > I think normally we bump runc by bumping cadvisor rather than on its own?
> 
> yes, that's traditionally been the main course of action.
> 
> I can help cut another cAdvisor release with updated runc, but I don't believe that's critical / has to block the update in k/k. I think we should be able to update runc by itself in k/k. @dims is that accurate/do you see any issues with that?

Correct. this update in this PR should be fine. 

/approve
/lgtm

/assign @liggitt

giuseppe commented:
> I can help cut another cAdvisor release with updated runc, but I don't believe that's critical / has to block the update in k/k. I think we should be able to update runc by itself in k/k. @dims is that accurate/do you see any issues with that?

I agree that this PR should not be blocked, as there are several fixes in libcontainer that are needed by the Kubelet.

Although there are some fixes in cAdvisor that are useful to get e2e tests to pass on cgroup v2.  Once we have a new cAdvisor release we can finally turn on tests for cgroup v2.

liggitt commented:
who on the node side can verify we are not impacted by the "potentially breaking changes" in rc94?

https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc94

> Potentially breaking changes:
> 
> cgroupv1: kernel memory limits are now always ignored, as kmemcg has been effectively deprecated by the kernel. Users should make use of regular memory cgroup controls. (opencontainers/runc#2840)
> libcontainer/cgroups: cgroup managers' Set now accept configs.Resources rather than configs.Cgroups (opencontainers/runc#2906)
> libcontainer/cgroups/systemd: reconnect and retry in case dbus connection is closed (after dbus restart) (opencontainers/runc#2923)
> libcontainer/cgroups/systemd: don't set limits in Apply (opencontainers/runc#2814)

SergeyKanzhelev commented:
> who on the node side can verify we are not impacted by the "potentially breaking changes" in rc94?

@bobbypage can you take a look?

/assign @bobbypage

bobbypage commented:
> cgroupv1: kernel memory limits are now always ignored, as kmemcg has been effectively deprecated by the kernel. Users should make use of regular memory cgroup controls. (opencontainers/runc#2840)

I don't thinks kubelet ever sets kernel memory limits, so this should not matter.

> libcontainer/cgroups: cgroup managers' Set now accept configs.Resources rather than configs.Cgroups (opencontainers/runc#2906)

This is API change that is being updated in this PR: https://github.com/kubernetes/kubernetes/pull/101888/commits/b49744f177087173d43d7771588d83b9df5a3c0b#diff-2f332b11d91d53583a6eeca75fc6e124aee6bc547a264e6987b89b1c786e0674

> libcontainer/cgroups/systemd: reconnect and retry in case dbus connection is closed (after dbus restart) (opencontainers/runc#2923)

This is bug fix, it should fix https://github.com/kubernetes/kubernetes/issues/100328

> libcontainer/cgroups/systemd: don't set limits in Apply (opencontainers/runc#2814)

It looks like we call `Apply` followed by `cgroupManagerImpl.Update` (which internally calls `Set`), so I don't think this should be issue.

https://github.com/kubernetes/kubernetes/blob/9739592e849c95f0cfc38cf9c1cbf241460096bb/pkg/kubelet/cm/cgroup_manager_linux.go#L644-L653

bobbypage commented:
/lgtm

liggitt commented:
/approve

SergeyKanzhelev commented:
/retest

odinuge commented on test/e2e_node/summary_test.go:
```diff
@@ -124,7 +124,7 @@ var _ = SIGDescribe("Summary API [NodeConformance]", func() {
 				"RSSBytes":        bounded(1*e2evolume.Kb, memoryLimit),
 				"PageFaults":      bounded(0, 1000000),
-				"MajorPageFaults": bounded(0, 10),
+				"MajorPageFaults": bounded(0, 1000),
```
We should revisit this, since this is changing a conformance test, and results in different behavior between cg v1 and v2. I don't have it on top of my head, but I think this will always be `0` on cg v1 since the metric only counts page faults on the pod cgroup itself.

odinuge reviewed (commented):
Sorry to be late at the game, but didn't see this PR before it marked some of my PRs as `Needs rebase`.... :/

With this version of runc, we need to do some refactoring to properly enforce cgroup values. Since runc changed they way they enforce limits (by not setting cgroup resources in Apply), this breaks our use of the systemd cgroup driver. As discussed in https://github.com/opencontainers/runc/issues/2813, a fix is needed in order to support it properly..

I have a PR to properly deal with systemd here: https://github.com/kubernetes/kubernetes/pull/98374.

I am however quite worried this PR (as in this PR, not mine) might cause regressions since both https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial and https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-conformance have been failing for quite some time. Those test suites are the only real ways to verify if these cause regressions. Did you run some of the tests locally (or on gce) @kolyshkin?

I think we should either try to merge the systemd-fix PR as soon as possible, and find some Googlers who can debug `node-kubelet-serial` tests so we can start run them again. I guess reverting this until we have working tests isn't an option, or?

liggitt commented:
The runc bump was referenced in a comment about a performance regression in 1.21 (https://github.com/kubernetes/kubernetes/issues/101989), so I expect there will be a desire to take this version back to the 1.21 release branch.

However, changing conformant behavior is a problem, so all the implications of this PR have to be understood and resolved before any backports to a release branch can be done.

@SergeyKanzhelev @bobbypage what would be your expectation here?
1. revert this PR, resolve the cgroup implications of the version bump, then bring it all in as a single update, then pick to 1.21?
2. add follow-up fixes, then pick all to 1.21?

Since this caused CI to turn red, my inclination is option 1.

liggitt commented:
revert open in https://github.com/kubernetes/kubernetes/pull/102094 to get CI running

odinuge commented:
Thanks for jumping in @liggitt !

> The runc bump was referenced in a comment about a performance regression in 1.21 (#101989), so I expect there will be a desire to take this version back to the 1.21 release branch.

Yeah, I think that was the plan yes. This is however quite a big dependency change, containing way more code and fixes than just the issue referenced here.

> However, changing conformant behavior is a problem, so all the implications of this PR have to be understood and resolved before any backports to a release branch can be done.

Yeah, that is a big issue. As far as I remember from working with these systemd-things last time, this change is effectively disabling all cgroup resource enforcement on pod (`kubepod-xyz`) and node level (`kubepods.slice`) (If I remember correctly, but it looks like that is the case.).

The change in the conformance test should also be done in a separate PR imo., so that all involved sigs understands how and why.

> 1. revert this PR, resolve the cgroup implications of the version bump, then bring it all in as a single update, then pick to 1.21?

I vote for this in short term, in order to avoid other devs having to debug these issues while working on other unrelated things. I do think we should get https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial up running again, and get proper signaling from that test suite in order to properly see that things works as expected.

liggitt commented:
> Also, the fixes required to mitigate these issues are quite complex, so I think it is a bit dangerous to rush them.

agreed. I asked in the issue about the performance regression how that can be resolved without taking rc94 back to release-1.21 (https://github.com/kubernetes/kubernetes/issues/101989#issuecomment-843177540)

smarterclayton commented:
Implication from

> Since runc changed they way they enforce limits (by not setting cgroup resources in Apply), this breaks our use of the systemd cgroup driver.

Is that if this was able to merge, that nothing in e2e is verifying the systemd-cgroup driver?  And that https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial is the way to get this going again?

If we can't guarantee that merges of runc don't break "resource management in kube", we definitely need that test (and others?) up and running ASAP.

odinuge commented:
> So roughly we're saying that nothing in presubmit e2e is verifying the systemd-cgroup driver? And that https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial is the way to get this going again but is currently broken?

Yeah, that's right as of now... But I 100% agree that it is a bad thing. `node-kubelet-serial` has gotten way too little love the last few years, and imo. it is the most important test suite we have (ofc. with my cgroup and sig-node bias).

I have been working for the last two years fixing failing tests in node-kubelet-serial, but since people usually don't look at it, it tends to fail, and issues and regressions are not properly detected. I have been busy with my master thesis for the last 7-8 months, and more and more of tests have started failing. When I have made runc bumps previously, I have always been running those tests myself, either locally or on my own gce account. Early in may, node-kubelet-serial started to go dark (as it is now), and I have not had time to dig more into the issue. Since it usually take a few hours to run it, it isn't suited for blocking pre submit. We could (and should) create an optional pre submit if we can get it to go green.

> If we can't guarantee that merges of runc don't break "resource management in kube", we definitely need that test (and others?) up and running ASAP, and potentially better ability to run those tests presubmit (as needed).

Yup! I have been using countless hours the last 2 years trying to keep it somewhat healthy, but I think some people with access to the gce account need to debug it as it is now.

We should also run the same test suite for cgroup v2 as well, otherwise I really don't see how we can graduate it to beta or GA with confidence that it _actually_ works as we expect it to do.

harche commented:
@smarterclayton https://github.com/kubernetes/kubernetes/issues/102092. We have node e2e jobs that verify that, but they aren't made blocking yet. We want to make them blocking, but we are still discussing if we should merge those crio jobs with existing docker job or make an independent blocking job, https://github.com/kubernetes/test-infra/pull/21278#issuecomment-843069486

odinuge commented:
> @harche:
> @smarterclayton #102092. We have node e2e jobs that verify that, but they aren't made blocking yet. We want to make them blocking, but we are still discussing if we should merge those crio jobs with existing docker job or make an independent blocking job, kubernetes/test-infra#21278 (comment)

Yeah, that test suite have some tests (a few) that will verify metrics etc. All tests that deal with cgroups etc. have to be executed in serial since they keep restarting kubelet etc., and those tests are all a part of `node-kubelet-serial`.

bobbypage commented:
Thanks @odinuge for raising concerns with the runc bump with and @liggitt for quick revert. It sounds like there's two issues so far with runc update that were found:

1. Breaking the `[sig-node] Summary API [NodeConformance]` test as described in https://github.com/kubernetes/kubernetes/issues/102092
2. Possible breaking change when applying limits when using systemd cgroup driver?

I'm not super clear on the second case. It looks like the current approach is we call [`Apply`](https://github.com/kubernetes/kubernetes/blob/9739592e849c95f0cfc38cf9c1cbf241460096bb/pkg/kubelet/cm/cgroup_manager_linux.go#L644-L653) followed by calling set on each subsystem with [`setSupportedSubsystemsV1`](https://github.com/kubernetes/kubernetes/blob/9739592e849c95f0cfc38cf9c1cbf241460096bb/pkg/kubelet/cm/cgroup_manager_linux.go#L373) to set the settings directly (i.e. not using libcontainer). Thus appears to me that kubelet does not depend on cgroup manager's `Set` which behavior was changed in runc rc94. Can you please clarify what is the regression caused by runc rc94 when combined with the systemd cgroup driver? Is there some issue with that approach that breaks when upgrading to runc rc94?

liggitt commented:
> Can you please clarify what is the regression caused by runc rc94 when combined with the systemd cgroup driver? Is there some issue with that approach that breaks when upgrading to runc rc94?

I'll defer to @odinuge (xref https://github.com/kubernetes/kubernetes/pull/101888#pullrequestreview-661762880)

odinuge commented:
> I'll defer to @odinuge (xref #101888 (review))

Thanks! Yeah, I think that comment describes it, but these systemd stuff is pretty complecated. This is a good resource: https://systemd.io/CGROUP_DELEGATION/

> @bobbypage: Possible breaking change when applying limits when using systemd cgroup driver?

With this PR, we never set systemd properties for resources like cpu and memory, we only write the cgroup files. Unless we use delegated subtrees (ref systemd docs), we should _not_ write to those files. Systemd will overwrite them.

Without this PR, we set systemd resources only at creation of systemd-slice/scope, that the PR I mentioned is fixing. https://github.com/kubernetes/kubernetes/pull/98374. Look at the issue that PR mark as "fixed" in order to understand the issue in current master.

To show this in real life, one can play with systemd like this:

```
$ sudo systemd-run -p MemoryAccounting=true -- sleep inf   
Running as unit: run-r6d7644f4a4054efe8a79bd296090fefc.service
$ cat /sys/fs/cgroup/system.slice/run-r6d7644f4a4054efe8a79bd296090fefc.service/memory.max 
max
$ sudo tee /sys/fs/cgroup/system.slice/run-r6d7644f4a4054efe8a79bd296090fefc.service/memory.max <<< 11111111111
11111111111
$ cat /sys/fs/cgroup/system.slice/run-r6d7644f4a4054efe8a79bd296090fefc.service/memory.max                     
11111108608
$ systemctl set-property run-r6d7644f4a4054efe8a79bd296090fefc.service MemoryAccounting=true # or some other random systemd thingy
$ cat /sys/fs/cgroup/system.slice/run-r6d7644f4a4054efe8a79bd296090fefc.service/memory.max  
max
```

This is certainly behavior we _don't_ would like to happen...

I guess I can write a book about this, but but this comment make it easier to understand the issue @liggitt @bobbypage @kolyshkin?

kolyshkin commented:
> I'm not super clear on the second case. It looks like the current approach is we call Apply followed by calling set on each subsystem with setSupportedSubsystemsV1 to set the settings directly (i.e. not using libcontainer cgroup manager). Thus appears to me that kubelet does not depend on cgroup manager's Set which behavior was changed in runc rc94. Can you please clarify what is the regression caused by runc rc94 when combined with the systemd cgroup driver? Is there some issue with that approach that breaks when upgrading to runc rc94?

`Set` behaviour has not changed, it's `Apply` behavior that was changed (in case systemd cgroup manager is used), details are at https://github.com/opencontainers/runc/pull/2814. I am still looking into the kubernetes e2e-node failures -- reproduced it locally but can't figure out why exactly it is failing. Going to check if https://github.com/kubernetes/kubernetes/pull/98374 helps.

kolyshkin commented:
> Breaking the [sig-node] Summary API [NodeConformance] test as described in #102092
1a. It looks like this is passing in containerd conformance test https://testgrid.k8s.io/sig-node-containerd#containerd-node-conformance&width=5 , but failing in crio conformance test https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-conformance

Spent a day on it, not sure what is causing the issue, will continue tomorrow.

odinuge commented:
> > Breaking the [sig-node] Summary API [NodeConformance] test as described in #102092
1a. It looks like this is passing in containerd conformance test https://testgrid.k8s.io/sig-node-containerd#containerd-node-conformance&width=5 , but failing in crio conformance test https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-conformance
> 
> Spent a day on it, not sure what is causing the issue, will continue tomorrow.

Are we talking about the same things, or are you looking at some other issues @kolyshkin?

I did some testing yesterday on my PR that fixes the systemd-stuff I mentioned in https://github.com/kubernetes/kubernetes/pull/101888#issuecomment-843496893 yesterday, and that does fix failing test @bobbypage mentiones.

See https://github.com/kubernetes/kubernetes/issues/102092#issuecomment-843136063 for more info.

Also see @harche's comment here: https://github.com/kubernetes/kubernetes/pull/98374#issuecomment-843853166

Here is a test run as pre submit: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98374/pull-kubernetes-node-crio-e2e/1394592396901093376/

kolyshkin commented:
@odinuge you're probably right; I have created https://github.com/kubernetes/kubernetes/pull/102147 that incorporates the change from #98374 to see what CI will show.
